src/protocol/sort/apply.rs

```diff
@@ -11,77 +12,121 @@ use permutation::Permutation;
 
 #[embed_doc_image("apply", "images/sort/apply.png")]
 #[embed_doc_image("apply_inv", "images/sort/apply_inv.png")]
+#[embed_doc_image("apply", "images/sort/apply.png")]
+#[embed_doc_image("apply_inv", "images/sort/apply_inv.png")]
+
 /// Permutation reorders (1, 2, . . . , m) into (σ(1), σ(2), . . . , σ(m)).
 /// For example, if σ(1) = 2, σ(2) = 3, σ(3) = 1, and σ(4) = 0, an input (A, B, C, D) is reordered into (C, D, B, A) by σ.
 /// ![Apply steps][apply]
-pub fn apply<T: Copy + Default, S>(mut permutation: Permutation, values: &mut S)
-where
-    S: AsMut<[T]>,
-{
-    permutation.apply_slice_in_place(values);
+pub fn apply<T: Copy + Default>(permutation: &[usize], values: &mut [T]) {
+    let mut permuted = bitvec![0; permutation.len()];
+    let mut tmp: T = T::default();
+
+    for i in 0..permutation.len() {
+        if permuted[i] == false {
+            mem::swap(&mut tmp, &mut values[i]);
+            let mut pos_i = i;
+            let mut pos_j = permutation[pos_i];
+            while pos_j != i {
+                values[pos_i] = values[pos_j];
```

Comment:
I think we could use `Vec::swap` here. That along with removing `mem::swap` and using local variable `tmp` will lead to `T: Clone` only

something like this should work
```rust
   for i in 0..permutation.len() {
        if !permuted[i] {
            let tmp = values[i].clone(); // still cheaper than mem::swap
            ...
            while pos_j != i {
                values.swap(pos_i, pos_j); // that does not require T:Copy
                ...
            }
            values[pos_i] = tmp; // destructive move, does not require T:Copy
        }
    }
```

Reply:
I'll give it a try!